### PR TITLE
IIS Webserver role

### PR DESCRIPTION
**Pull request by mcka1n**
### Changes

* Adding a new Role `role::webserver` to manage Windows IIS servers
   * This role uses `profile::platform::windows::webserver::iis_server` which installs and configures IIS.
* Reboots the Windows machine after Chocolatey installs 7zip. (via `puppetlabs-reboot`)

----

Note for the reviewer: 

This is how I verified that after 7Zip was uninstalled, the next time Puppet will enforce the installation again and reboot after that.

image

### CODE REVIEW

File: site-modules/profile/manifests/platform/windows/webserver/iis_server.pp [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    extract_path => 'C:/inetpub/minimal',
+    cleanup      => true,
+    notify  => Exec['MoveIndex'],
+  }
+
+  exec { 'MoveIndex':
+    command     => "Copy-Item C:/inetpub/minimal/af0e24303d241b888152bd1cd7c9063d-ad273bebc01c6dac176da7a5f3c38c4d9a584521/index.html C:/inetpub/minimal/index.html",
+    provider    => powershell,
+    refreshonly => true,
+    require     => Archive['C:/inetpub/minimal/af0e24303d241b888152bd1cd7c9063d']
```
Comment (dylanratcliffe) [inline site-modules/profile/manifests/platform/windows/webserver/iis_server.pp]: This `require` param isn't required (no pun intended). A `notify` implies `before` meaning that the ordering relationship `Archive['C:/inetpub/minimal/af0e24303d241b888152bd1cd7c9063d']` comes before `Exec['MoveIndex']` already exists and specifying a require from this direction doesn't actually do anything that isn't already there

**Review by dylanratcliffe — CHANGES_REQUESTED** [review]
I've added one comment about ordering but I suspect this works perfectly in its current form. However there are a few things to consider that might improve the functionality:

* Can we get rid of the download-then-move mechanism? Seems a but messy
* What would happen if someone manually changed the `index.html`? Would Puppet fix it?
* Do we need to ensure permissions on the `index.html` once it's created? Would it need to be owned by some kind of IIS user in order for IIS to read it? We should presumably at least make sure that it isn't world writable for example as we don't want people going in and changing it

Let me know what you think of the above

File: site-modules/profile/manifests/platform/windows/webserver/iis_server.pp (hunk 1 above)
Comment (mcka1n) [reply site-modules/profile/manifests/platform/windows/webserver/iis_server.pp]: Makes total sense Dylan!

**Comment by mcka1n** [comment]
> I've added one comment about ordering but I suspect this works perfectly in its current form. However there are a few things to consider that might improve the functionality:
> 
> * Can we get rid of the download-then-move mechanism? Seems a but messy
> * What would happen if someone manually changed the `index.html`? Would Puppet fix it?
> * Do we need to ensure permissions on the `index.html` once it's created? Would it need to be owned by some kind of IIS user in order for IIS to read it? We should presumably at least make sure that it isn't world writable for example as we don't want people going in and changing it
> 
> Let me know what you think of the above

Hey @dylanratcliffe 

So I made some changes based on your suggestions.

- I removed the "download-then-move" logic, I switched to a simpler version with a `File` resource
- I'm changing the permissions so everyone can `read` but not make changes on the `index.html`
- The group `IIS_IUSRS` was used to allow IIS to read and render the HTML on port 80

### Test

To test that no one else could modify the HTML file, I went to the Windows server and tried to edit it with VIM:

image

Let me know what you think :)

**Review by dylanratcliffe — APPROVED** [review]
Looking great

**Pull request merged**